### DISCUSSION TITLE
How to convert different Filecoin addresses programmatically?

### DISCUSSION THREAD

**Discussion by longfeiWan9** [filecoin-project/community#650] [Q&A]
Filecoin address types:
- f0 - ID address for any address on Filecoin 
- f1 - secp256k1 wallet address
- f2 - actor address
- f3 - BLS wallet address
- f410 - Ethereum-compatible address

There are a couple of libraries you can use for the address conversion.
- [glif/filecoin-address](https://github.com/glifio/modules/tree/primary/packages/filecoin-address)
- [izari-filecoin](https://github.com/Zondax/izari-filecoin/tree/master)

#### f410 <-> 0x
1. 0x -> f410
we are going to use [glif/filecoin-address](https://github.com/glifio/modules/tree/primary/packages/filecoin-address) library. So you need to install it into your project first. 
`npm install @glif/filecoin-address`
Then you can write code to convert your eth address to Filecoin f410 address.
    ```
    const {newDelegatedEthAddress} = require('@glif/filecoin-address')

    const filAddress = newDelegatedEthAddress('0x44061AA8Df5b33a997CE97d80c700d0C655Dc3f2');
    console.log(filAddress.toString());
    ```
2. f410 -> 0x
we can also use  [glif/filecoin-address](https://github.com/glifio/modules/tree/primary/packages/filecoin-address) library. 
     ```
    const {ethAddressFromDelegated} = require('@glif/filecoin-address')

    const ethAddress = ethAddressFromDelegated('f410f2oekwcmo2pueydmaq53eic2i62crtbeyuzx2gmy');
    console.log(ethAddress.toString());
    ```
#### f0 -> 0x
All Filecoin addresses (f1, f2, f3) will have a unique ID address assigned to it, and this ID address will be finalized after 900 epochs.  ID address can be converted to 0x address which could be used to send FIL to your other Filecoin wallet using Ethereum-compatible tools.
```
const ethIDAddress = ethAddressFromID('t01073');
console.log(ethIDAddress.toString());
// output: 0xff00000000000000000000000000000000000431
```
#### f1/f2/f3 <-> f410 or 0x
You cannot.

**Comment by Stebalien** [comment]
Anyone reading this should also read https://github.com/glifio/modules/issues/318. There isn't a 1:1 mapping between `0x` and `f4`, so the instructions here aren't quite correct.

**Comment by longfeiWan9** [comment]
A new JSON-RPC API `FilecoinAddressToEthAddress` was introduced with Lotus V1.29.0 releases to enable querying 0x address for any types of Filecoin addresses. 
- f410 -> 0x 
   `t410fghja63ghyzd54cbswfapglsutx5e64xnm7mazwi` -> `0x31d20f6cc7c647de0832b140f32e549dfa4f72ed` 
- f1/f2/f2: will return 0x address for their corresponding f0 ID address.
    `f1alg2sxw32ns3ech2w7r3dmp2gl2fputkl7x7jta ` -> `0xff00000000000000000000000000000000000064`

You can learn how to use this API [here](https://github.com/filecoin-project/lotus/blob/master/documentation/en/api-v1-unstable-methods.md#filecoinaddresstoethaddress).

**Comment by Pfed-prog** [comment]
My further understanding is that there is no bridge between f410 and f1
I am unable to withdraw funds from 0x wallet to bybit or kucoin despite using forwarders
The only working solution for me was Binance
